var/spack/repos/builtin/packages/fftx/package.py

```diff
@@ -0,0 +1,92 @@
+# Copyright 2013-2022 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+import subprocess
+
+from spack import *
+
+
+class Fftx(CMakePackage):
+    """FFTX is the exascale follow-on to the FFTW open source discrete FFT
+    package for executing the Fast Fourier Transform as well as higher-level
+    operations composed of linear operations combined with DFT transforms."""
+
+    homepage = "https://spiral.net"
+    url      = "https://github.com/spiral-software/fftx/archive/0.9.0.tar.gz"
+    git      = "https://github.com/spiral-software/fftx.git"
+
+    maintainers = ['spiralgen']
+
+    version('develop', branch='develop')
+    version('main',  branch='main')
+    version('0.9.0', sha256='d4930e9b959fd56bb63543b359ca09a7ae2c56db3ffc28d7d3628d92fc79ce12')
+
+    variant('build_type', default='Release',
+            values=('Debug', 'Release', 'RelWithDebInfo', 'MinSizeRel'),
+            description='The build type to build')
+
+    depends_on('spiral-software')
+    #  depends_on('spiral-package-fftx')
+    #  depends_on('spiral-package-simt')
+    #  depends_on('spiral-package-mpi')
+
+    @run_before('cmake')
+    def create_lib_source_code(self):
+        #  From directory examples/library run the build-lib-code.sh script
+        with working_dir(join_path(self.stage.source_path, 'examples/library')):
+            bld_script = join_path(self.stage.source_path,
+                                   'examples/library/build-lib-code.sh')
+            result = subprocess.run(bld_script, shell=True, check=False)
+            res = result.returncode
+            print(result)
+            #  Use the script result to set the build parameter for use with cmake
+            if res == 0:
+                self.build_config = '-D_codegen=CPU'
+            elif res == 1:
+                self.build_config = '-D_codegen=CUDA'
+            elif res == 2:
+                self.build_config = '-D_codegen=HIP'
```

Comment:
As I mention elsewhere, refer to Inheriting from `CudaPackage` gives you two variants: `cuda` (https://github.com/spack/spack/blob/3c0e422509303703d87baaf0a9121df8376881ea/lib/spack/spack/build_systems/cuda.py#L34) and `cuda_arch` and a bunch of `conflicts` you can get "for free".

Similarly, `ROCmPackage` gives you `rocm` (https://github.com/spack/spack/blob/3c0e422509303703d87baaf0a9121df8376881ea/lib/spack/spack/build_systems/rocm.py#L98) and `amd_gpu` plus conflicts.

If your package doesn't support both during the same build, then you'll want to add a conflict, such as:

https://github.com/spack/spack/blob/3c0e422509303703d87baaf0a9121df8376881ea/var/spack/repos/builtin/packages/slate/package.py#L56 

for an example of checking variants and setting an option for `cmake_args`.

Reply:
Using variants